### PR TITLE
Fix compiler error with complex instanciation and misc improvements (HAVE_MPI -> HAVE_IFPACK2_MPI)

### LINKED ISSUE #2871: Ifpack2: BlockTriDiContainer calls unknown method
@trilinos/ifpack2 
@kyungjoo-kim @ambrad 

## Expectations
Ifpack2 should build...

## Current Behavior
If does not build, gcc returns the following error:

```
In file included from /home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainer.cpp:59:0:
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainerUtil.hpp: In instantiation of ‘static tif_utest::BlockTriDiContainerTester<Scalar, LO, GO>::Int tif_utest::BlockTriDiContainerTester<Scalar, LO, GO>::test_BR_BTDC(const Teuchos::RCP<const Teuchos::Comm<int> >&, const StructuredBlock&, const StructuredBlockPart&, tif_utest::BlockTriDiContainerTester<Scalar, LO, GO>::Int, tif_utest::BlockTriDiContainerTester<Scalar, LO, GO>::Int, bool, bool, bool, bool, bool, const string&) [with Scalar = std::complex<double>; LO = int; GO = int; tif_utest::BlockTriDiContainerTester<Scalar, LO, GO>::Int = int; tif_utest::BlockTriDiContainerTester<Scalar, LO, GO>::StructuredBlock = tif_utest::BlockCrsMatrixMaker<std::complex<double>, int, int>::StructuredBlock; tif_utest::BlockTriDiContainerTester<Scalar, LO, GO>::StructuredBlockPart = tif_utest::BlockCrsMatrixMaker<std::complex<double>, int, int>::StructuredBlockPart; std::string = std::basic_string<char>]’:
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainer.cpp:292:51:   required from ‘LO run_teuchos_tests(const Input&, Teuchos::FancyOStream&, bool&) [with Scalar = std::complex<double>; LO = int; GO = int; Teuchos::FancyOStream = Teuchos::basic_FancyOStream<char>]’
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainer.cpp:313:53:   required from ‘void Ifpack2BlockTriDi_Unit_UnitTest<Scalar, LO, GO>::runUnitTestImpl(Teuchos::FancyOStream&, bool&) const [with Scalar = std::complex<double>; LO = int; GO = int; Teuchos::FancyOStream = Teuchos::basic_FancyOStream<char>]’
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainer.cpp:321:1:   required from here
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainerUtil.hpp:238:9: error: ‘class Ifpack2::BlockTriDiContainer<Tpetra::RowMatrix<std::complex<double>, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Ifpack2::BlockTriDiContainerDetails::ImplNotAvailTag>’ has no member named ‘createDefaultComputeParameters’
         T_bare->compute(T_bare->createDefaultComputeParameters());
         ^
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainerUtil.hpp:293:49: error: ‘class Ifpack2::BlockTriDiContainer<Tpetra::RowMatrix<std::complex<double>, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Ifpack2::BlockTriDiContainerDetails::ImplNotAvailTag>’ has no member named ‘getNorms0’
               const auto n0 = T_bare->getNorms0();
                                                 ^
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainerUtil.hpp:294:53: error: ‘class Ifpack2::BlockTriDiContainer<Tpetra::RowMatrix<std::complex<double>, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Ifpack2::BlockTriDiContainerDetails::ImplNotAvailTag>’ has no member named ‘getNormsFinal’
               const auto nf = T_bare->getNormsFinal();
                                                     ^
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainerUtil.hpp:311:63: error: ‘class Ifpack2::BlockTriDiContainer<Tpetra::RowMatrix<std::complex<double>, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Ifpack2::BlockTriDiContainerDetails::ImplNotAvailTag>’ has no member named ‘createDefaultApplyParameters’
             auto input = T_bare->createDefaultApplyParameters();
                                                               ^
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainerUtil.hpp:318:19: error: ‘class Ifpack2::BlockTriDiContainer<Tpetra::RowMatrix<std::complex<double>, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Ifpack2::BlockTriDiContainerDetails::ImplNotAvailTag>’ has no member named ‘createDefaultApplyParameters’
             input = T_bare->createDefaultApplyParameters();
                   ^
/home/lberge/Research/Trilinos_lucbv/packages/ifpack2/test/unit_tests/Ifpack2_UnitTestBlockTriDiContainerUtil.hpp:325:21: error: ‘class Ifpack2::BlockTriDiContainer<Tpetra::RowMatrix<std::complex<double>, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >, Ifpack2::BlockTriDiContainerDetails::ImplNotAvailTag>’ has no member named ‘createDefaultApplyParameters’
               input = T_bare->createDefaultApplyParameters();
```

## Motivation and Context
This is blocking development and it also prevent nightly tests to run properly since Ifpack2 gets disabled by CMake.

## Definition of Done
Ifpack2 can be build again

## Possible Solution
It looks from the template: `Ifpack2::BlockTriDiContainerDetails::ImplNotAvailTag` that some implementation might be missing in which case BlockTriDiContatiner might not be ready to be pushed in develop?
Or the missing implementation needs to be added or guarded to avoid compilation issues?

## Steps to Reproduce
Get a fresh version of trilinos and enable Ifpack2, I tried with and without `openmp`, both are failing.

## Your Environment
- Trilinos SHA1: 0d2d3c9
- source Trilinos/cmake/load_sems_dev_env.sh
- Trilinos_ENABLE_Ifpack2

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
This reverts PR #2875 which is accidently merged.

### CODE REVIEW

File: packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp [hunk 1]
```diff
@@ -50,6 +50,13 @@
 
 #include <Kokkos_ArithTraits.hpp>
 #include <KokkosBatched_Util.hpp>
+
+/// complex is_convertible override
+namespace std {
+  template<typename T1, typename T2> struct is_convertible<Kokkos::complex<T1>, T2> { static constexpr bool value = std::is_convertible<T1,T2>::value; };
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp]: It's forbidden to override `std::is_convertible`: https://github.com/kokkos/kokkos/issues/1658#issuecomment-394508562

It should also be unnecessary.

Comment (kyungjoo-kim) [reply packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp]: Then, this needs to wait until the next kokkos kernels promotion. I put the static_assert in kokkoskernels in SIMD constructor. @lucbv

Comment (ndellingwood) [reply packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp]: @kyungjoo-kim submitted PR kokkos/kokkos-kernels#251 as a fix for the `is_convertible` troubles; following spot-check, a matching patch to Trilinos can be issued as a hotfix. 
For bookkeeping may be nicer to issue the PR separately from this one, but that increases autotester time so may be more efficient to just include it here.

**Comment by kyungjoo-kim** [comment]
Now this PR does not include fix for issue #2871 but includes some improvement
- paramter struct is nested in the blocktridicontainer.
- HAVE_MPI is replaced with HAVE_IFPACK2_MPI

This fix will be submitted in another PR after kokkoskernels is updated.

**Comment by ndellingwood** [comment]
@mhoemmen the PR has been updated, can you update your review when you get a chance? Thanks!

**Comment by mhoemmen** [comment]
FYI, Ifpack2 doesn't actually define `HAVE_IFPACK2_MPI`.  You'll have to make sure the option gets set correctly in `ifpack2/CMakeLists.txt`.

**Comment by mhoemmen** [comment]
You could just use `HAVE_TEUCHOS_MPI` which should already be defined.

**Comment by ndellingwood** [comment]
The kokkos-kernels changes passed spot-check on apollo, see kokkos/kokkos-kernels#251

**Comment by kyungjoo-kim** [comment]
Oh.... really.... 

@mhoemmen I just saw Ifpack2_config.h.in and found ``#cmakedefine HAVE_IFPACK2_MPI``. So, I thought that somebody already put necessary define the cmake option but it is not. Thanks for checking them. This is so frustrating...

**Comment by mhoemmen** [comment]
@kyungjoo-kim wrote:
> I just saw `Ifpack2_config.h.in` and found `#cmakedefine HAVE_IFPACK2_MPI`. So, I thought that somebody already put necessary define the cmake option but it is not. Thanks for checking them. This is so frustrating...

Welcome to my career ;-)

FYI, the `#cmakedefine` is not enough; the (sub)package's `CMakeLists.txt` file must define a corresponding CMake option, via either `TRIBITS_ADD_OPTION_AND_DEFINE` or `GLOBAL_SET`.

**Comment by mhoemmen** [comment]
@kyungjoo-kim Hm, the Ifpack2 BlockTriDiag test is failing, on lines 300 and 310:

https://testing-vm.sandia.gov/cdash/testDetails.php?test=47931832&build=3580142

**Comment by kyungjoo-kim** [comment]
Let's see how the others are tested.

**Comment by kyungjoo-kim** [comment]
@mhoemmen I tested the intel build with the same sems environment but it passes on my workstation. As the error only happens in overlap comm/computation, I hope that this is just one time network issue. After this PR test is done, I will redo the test again to check it out.

**Comment by kyungjoo-kim** [comment]
@mhoemmen @lucbv @ndellingwood It passed all tests. Please update your review and let's finish this PR.

**Pull request merged**